### PR TITLE
[lldb] Consider bound types when deciding if the overall type is dynamic

### PR DESCRIPTION
**Pull request by augusto2112**

### CODE REVIEW

**Comment by augusto2112** [comment]
@swift-ci test

File: lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp [hunk 1]
```diff
@@ -1990,37 +1990,59 @@ bool TypeSystemSwiftTypeRef::IsPossibleDynamicType(opaque_compiler_type_t type,
...
+      case Node::Kind::DynamicSelf:
+        return true;
+      case Node::Kind::BoundGenericStructure:
+      case Node::Kind::BoundGenericEnum: {
+        if (node->getNumChildren() < 2)
+          return false;
+        NodePointer type_list = node->getLastChild();
+        if (type_list->getKind() != Node::Kind::TypeList)
+          return false;
+        for (size_t i = 0; i < type_list->getNumChildren(); ++i) {
```
Comment (adrian-prantl) [inline lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp:2021]: `for (NodePointer child : *type_list) {`

File: lldb/test/Shell/SwiftREPL/Dict.test [hunk 2]
```diff
@@ -7,19 +7,22 @@
 // once per element.
 
 [1:2, 2:3, 3:9]
-// DICT-LABEL: $R0: [Int : Int] = 3 key/value pairs {
+// DICT-LABEL: $R0: [Int : Int] =
```
Comment (adrian-prantl) [inline lldb/test/Shell/SwiftREPL/Dict.test]: I'm just curious: what does the full new output look like? Is the `3 key/value pairs` on a different line now?

**Review by adrian-prantl — APPROVED** [review]
Looks good (minor stylistic comment inline)

File: lldb/test/Shell/SwiftREPL/Dict.test (hunk 2 above)
Comment (augusto2112) [reply lldb/test/Shell/SwiftREPL/Dict.test]: Yes, the new output looks like this: 
```
x: [Foundation.URL : Int] = {
  _variant = {
    object = {
      rawValue = 2 key/value pairs {
        [0] = {
          key = "https://github.com"
          value = 4
        }
        [1] = {
          key = "https://apple.com"
          value = 23
        }
      }
    }
  }
}
```

Comment (adrian-prantl) [reply lldb/test/Shell/SwiftREPL/Dict.test]: Shouldn't the `_variant = ` be hidden by the Dictionary data formatter? Is the formatter not triggering?

Comment (augusto2112) [reply lldb/test/Shell/SwiftREPL/Dict.test]: @adrian-prantl an update on this: the changed isn't caused by this patch, but by implementing `TypeSystemSwiftTypeRef::GetFullyUnqualifiedType`, which this was in preparation for. I've removed the test changes and will look into why a typeref type ins't triggering the formatter.

**Comment by augusto2112** [comment]
@swift-ci test

File: lldb/test/Shell/SwiftREPL/Dict.test (hunk 2 above)
Comment (adrian-prantl) [reply lldb/test/Shell/SwiftREPL/Dict.test]: Thanks for investigating!

**Pull request merged**